src/System.IO.Pipelines/ReadableBufferReader.cs

```diff
@@ -53,78 +54,107 @@ public ReadCursor Cursor
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Peek()
         {
-            if (_end)
+            if (!_end)
             {
-                return -1;
+                return _currentSpan[_index];
             }
-            return _currentSpan[_index];
+
+            return -1;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Take()
         {
-            if (_end)
+            if (!_end)
```

Comment:
Why would the JIT do it, how would it know to do it? You'd need to do PGO runs.

It does for exceptions (and correctly constructed throw helpers) as they should be uncommon paths; i.e. it moves the actual exception `throw ...` or call to `void Throw() -> throw` to end of method so if the "is invalid" test passes it is conditional jump forward (non predicted) to call the exception code.

It does for while and for loops, where possible; turning them into do while loops, so the test is at the end and its a conditional jump backwards (predicted) from the test. Backwards jump are predicted as they are loops and loops normally loop.

Ternary operators you want to put the more likely item first e.g. `var value = condition ? likelyResult : unlikelyResult` as it gets turned into:

```csharp
if (condition)
{
    value = likelyResult;
}
else
{ 
    value = unlikelyResult;
}
```
which becomes
```asm
test condition
jmpNotTrue second ; jmp forward not predicted
    value = likelyResult 
    jmp end
second:
    value = unlikelyResult
end:
```

...however the C# compiler on the other hand will reverse this code and change it to 
```csharp
if (_end)
{
   return -1;
}
...
```
```
.method public hidebysig instance int32 
    Take() cil managed 
  {
    .maxstack 4

    // [66 13 - 66 22]
    IL_0000: ldarg.0      // this
    IL_0001: ldfld        bool System.IO.Pipelines.ReadableBufferReader::_end
    IL_0006: brfalse.s    IL_000a

    // [68 17 - 68 27]
    IL_0008: ldc.i4.m1    
    IL_0009: ret          
```
defeating the change and deoptimizing the code; so I'd consider that a bug in csharp compiler